question-servers/freeformPythonLib/prairielearn.py

```diff
@@ -17,6 +18,31 @@
 import collections
 
 
+# TODO in the future it would be nice to have a way to use the `Final` type
+# around some of these values for the different functions so that the type
+# checker knows which functions are allowed to modify which values.
+# This may require subclassing QuestionData, and constructing the `data`
+# dictionary through class constructors instead of the usual was
+class QuestionData(TypedDict):
+    params: Dict[str, Any]
+    correct_answers: Dict[str, Any]
+    submitted_answers: Dict[str, Any]
+    format_errors: Dict[str, Any]
+    partial_scores: Dict[str, Any]
+    score: float
+    feedback: Dict[str, Any]
+    variant_seed: int
+    options: Dict[str, Any]
+    raw_submitted_answers: Dict[str, Any]
+    editable: bool
+    panel: Literal['question', 'submission', 'answer']
+    extensions: Dict[str, Any]
+
+
+class ElementTestData(QuestionData):
+    test_type: Literal['correct', 'incorrect', 'invalid']
+
+
```

Comment:
I'm not sure if this file is the best

I also don't like how many `Any`s there are here, there are a few different things we could do about it, but I'm not convinced which one is best yet. 

- leave them as `Any`: don't have to write as much code, but less type safe
- change them to `unknown`: individual elements will need to do explicit run time checking or casts in order to access them, but then it will be type safe
- Have individual elements make a new type inheriting from this one that declares their actual type for their use case: more code, may or may not play together well with inheriting from this type to make certain things `Final` we could end up with a complicated inheritance structure. That could be good though if it gives us really nice type safety everywhere

Reply:
Oh wait, it seems like types of properties can't be redefined in subclasses, even if the new type is a subtype of the type in the parent class. That is really annoying. I guess that means that if we want to mark certain things as `Final` depending on which function we are calling (e.g. `options` should be `Final` in the `prepare` function because it can't be edited there), or if we want to more specific types for the `Any`s, we might have to make entirely new classes instead of just inheriting. That's very annoying.

Reply:
> we might have to make entirely new classes instead of just inheriting. That's very annoying.

IMO this would be completely acceptable. These types virtually never change, so I'm not concerned about duplicating a bit of code.